agent/command/s3_copy.go

```diff
@@ -75,105 +95,140 @@ type s3Loc struct {
-func (c *s3copy) Name() string { return "s3Copy.copy" }
+func s3CopyFactory() Command      { return &s3copy{} }
+func (s3pc *s3copy) Name() string { return "s3Copy.copy" }
 
-// ParseParams decodes the S3 push command parameters that are
-// specified as part of an S3CopyPlugin command; this is required
-// to satisfy the 'Command' interface
-func (c *s3copy) ParseParams(params map[string]interface{}) error {
-	if err := mapstructure.Decode(params, c); err != nil {
-		return errors.Wrapf(err, "error decoding %v params", c.Name())
+// ParseParams decodes the S3 push command parameters
+func (s3pc *s3copy) ParseParams(params map[string]interface{}) error {
+	decoder, err := mapstructure.NewDecoder(&mapstructure.DecoderConfig{
+		WeaklyTypedInput: true,
+		Result:           s3pc,
+	})
+	if err != nil {
+		return errors.WithStack(err)
 	}
-	if err := c.validateParams(); err != nil {
-		return errors.Wrapf(err, "error validating %v params", c.Name())
+
+	if err := decoder.Decode(params); err != nil {
+		return errors.Wrapf(err, "error decoding %s params", s3pc.Name())
 	}
-	return nil
+
+	return s3pc.validate()
 }
 
 // validateParams is a helper function that ensures all
 // the fields necessary for carrying out an S3 copy operation are present
-func (c *s3copy) validateParams() error {
-	if c.AwsKey == "" {
-		return errors.New("s3 AWS key cannot be blank")
+func (s3pc *s3copy) validate() error {
+	catcher := grip.NewSimpleCatcher()
+
+	// make sure the command params are valid
+	if s3pc.AwsKey == "" {
+		catcher.Add(errors.New("aws_key cannot be blank"))
 	}
-	if c.AwsSecret == "" {
-		return errors.New("s3 AWS secret cannot be blank")
+	if s3pc.AwsSecret == "" {
+		catcher.Add(errors.New("aws_secret cannot be blank"))
 	}
-	for _, s3CopyFile := range c.S3CopyFiles {
+
+	for _, s3CopyFile := range s3pc.S3CopyFiles {
+		// make sure both buckets are valid
+		if err := validateS3BucketName(s3CopyFile.Source.Bucket); err != nil {
+			return errors.Wrapf(err, "%v is an invalid bucket name", s3CopyFile.Source.Bucket)
+		}
+		if err := validateS3BucketName(s3CopyFile.Destination.Bucket); err != nil {
+			return errors.Wrapf(err, "%v is an invalid bucket name", s3CopyFile.Destination.Bucket)
+		}
+
 		if s3CopyFile.Source.Region == "" {
 			s3CopyFile.Source.Region = endpoints.UsEast1RegionID
 		}
 		if s3CopyFile.Destination.Region == "" {
 			s3CopyFile.Destination.Region = endpoints.UsEast1RegionID
 		}
-		if s3CopyFile.Source.Bucket == "" {
-			return errors.New("s3 source bucket cannot be blank")
-		}
-		if s3CopyFile.Destination.Bucket == "" {
-			return errors.New("s3 destination bucket cannot be blank")
-		}
-		if s3CopyFile.Source.Path == "" {
-			return errors.New("s3 source path cannot be blank")
-		}
-		if s3CopyFile.Destination.Path == "" {
-			return errors.New("s3 destination path cannot be blank")
-		}
-		if s3CopyFile.Permissions == "" {
-			s3CopyFile.Permissions = s3.BucketCannedACLPublicRead
```

Comment:
I think we should still be defaulting permissions in case current commands rely on this.